Genes pipeline

afg1 opened the pull request:
This PR merges the genes pipeline into dev so we have a stable base to continue developing on, especially for search export

afg1 committed:
Code to implement the core random forect classification workflow

afg1 committed:
Add new random forest gene inferencing to the CLI

afg1 committed:
Nextflow workflow for genes

afg1 committed:
Update genes config with model remotes

afg1 committed:
Update poetry for new dependencies

afg1 committed:
Add entry workflow

afg1 committed:
Include taxa query

afg1 committed:
nextflow bugfixing

afg1 committed:
run genes workflow

afg1 committed:
Remove duplicated preprocessing function

afg1 committed:
Remove more duplication

afg1 committed:
Update query to use active mapped regions view

This pulls less data, but is probably more correct

afg1 committed:
Fix a couple of bugs with using a module of the same name as click command

afg1 committed:
Switch to using onnxruntime for inference to avoid sklearn versioning problems

afg1 committed:
Add onnxruntime dependency for inference

afg1 committed:
Require single threaded execution to avoid SLURM thread affinity peroblems

afg1 committed:
Have to write JSON format since members is a list of regions, so CSV won't work

afg1 committed:
Add batching to inference code

Should save memory on massive datasets

afg1 committed:
Change output sought to JSON from CSV

afg1 committed:
Fix indentation scope error

afg1 committed:
Early return empty dataframe if there are no features to work with

afg1 committed:
Handle empty results earlier to avoid crashing

taxid 8839 apparently has no active mapped regions

afg1 committed:
Explicitly set SLURM CPU count from environment

afg1 committed:
Optimize feature generation

afg1 committed:
Merge in GFF to transcript parquet conversion function

afg1 committed:
Fix a missing pybedtools dependency

afg1 committed:
Fix a missing import in genes data

afg1 committed:
Multiprocessing optimisation

afg1 committed:
Multiprocessing related cli arguments

afg1 committed:
Improve assembly name extraction

afg1 committed:
Merge branch 'master' into genes-pipeline

afg1 committed:
Fix poetry lock file after merge

afg1 committed:
Update to use correct SLURM environment variable

afg1 committed:
Pass session options to ONNX runtime properly

afg1 committed:
Single load node2vec SO model

afg1 committed:
Get assembly ID from database, given region name

afg1 committed:
Fix some potential unbound variable problems

afg1 committed:
Look up assembly IDs from the database during gff conversion

afg1 committed:
Pass database connection string through for assembly_id lookup

afg1 committed:
Fix import path for assembly id lookup

afg1 committed:
Fix a syntax error in the import

afg1 committed:
Fix import path again

afg1 committed:
Explicit naming of transcripts file generated from GFF

So it is easy to distinguish from other parquet we will be generating

afg1 committed:
Fix output path for transcripts

Suffix was invalid

afg1 committed:
Add gene bedfile export code

afg1 committed:
Group feature calculation by strand as well

I think some opposite strand transcripts were leading to 'bridging' comparisons in feature calculation, which then result in overzealous grouping of transcripts. This seems to remove the problem in the spot checks I've done

afg1 committed:
Add gene release merging code

Uses perfect overlap as a first filter to merge genes whose regions have not changed (but maybe picked up members), and then uses region overlap >0.9 for genes within 1kb of each other

Discards genes from previous releases that are not present in the current/next release

afg1 committed:
Optimised preprocessing code

Much faster for the MT chromosome especially. Has been profiled, I don't think there's much to gain now

afg1 committed:
Clean up old code

afg1 committed:
Import missing overlap calculation

afg1 committed:
Remove some unused CLI stuff from the old method

afg1 committed:
Fix module path

afg1 committed:
CLI and code for upsert into the database

blakesweeney committed:
Outline initial gene export work

afg1 committed:
Only search for candidates with similar start locations, don't consider stop locations as well

afg1 committed:
Add filtering to ignore piRNAs

afg1 committed:
Fix capitalisation for expression atlas DESCR string

afg1 committed:
Handle metadata for getes, including descriptions and types

Uses a score-based selection method derived from the same kind of logic in precompute. Works ok on human genes, not yet tested on other organisms

afg1 committed:
Add CLI for metadata related functions

afg1 committed:
Query for genes model building training data extraction

Extracts the needed data from our records of ensembl genes for human only

afg1 committed:
Process full dataframe

afg1 committed:
Use unique urs_taxids in retrieval

afg1 committed:
Fix query for cm hits

afg1 committed:
Use cursor context manager

afg1 committed:
Add explicit commit before closing connection

Keeps getting stuck in indle in transaction state, this may fix it

afg1 committed:
Count versions from 1 instead of zero

afg1 committed:
Make sure accessions fetching properly terminates connection

afg1 committed:
Add error handling around connections for fetching accessions and cm hits

afg1 committed:
Split union query in two and concatenate in polars

afg1 committed:
Better creation of empty dataframes for concatenation

afg1 committed:
Use relaxed concatenation and fix a tuple error

afg1 committed:
Remove erroneous commits

afg1 committed:
Try explicitly closing the cursor in multi query functions

afg1 committed:
Add logging

afg1 committed:
Track cm hit concatenation

afg1 committed:
Relax concatenation with accessions

afg1 committed:
Use from_dict with explicit schema

afg1 committed:
RealDictCursor returns an iterable of dicts, so ude from_dicts to convert ti polars

afg1 committed:
Use closures and close connection after each query

Also do a better job with scoping in connections

afg1 committed:
Turn off multithreaded sorting

afg1 committed:
Use spawn context for multithreading to play nicely with polars

afg1 committed:
Merge in training code from old repo

Includes training specific data handling, e.g. fetching gene coords from ensembl and doing data splits

afg1 committed:
Make step-specific subgroups in the genes CLI

afg1 committed:
Add training options to the CLI

afg1 committed:
Add skl2onnx for model conversion

afg1 committed:
Update pipeline container to latest debian base image

Also updates python minor version from 3.11.0 to 3.11.13. Hopefully that won't break anything...

afg1 committed:
Upgrade some obsoleted packages

afg1 committed:
Fix misspelled apt argument

afg1 committed:
Try upgrading to blat37

afg1 committed:
Fix blat url

afg1 committed:
Inline edit to the blat makefile to skip building webBlat, which is incompatible with libssl3

afg1 committed:
Try building ONLY blat by cdin to the blat directory and making there

This assumes all blat dependencies are in the blat subdirectory and that no interdependencies exist between folders

afg1 committed:
Build blat dependencies in lib

afg1 committed:
Directly edit makefile to preserve dependency tracking across targets

afg1 committed:
Revert to building everything for blat

afg1 committed:
Switch to using precompiled blat executables

afg1 committed:
FIx metadata storage

Selects the right columns in the right order for the insert query

afg1 committed:
Code to run rf model evaluation

afg1 committed:
Add evaluation option for genes model

afg1 committed:
Add datasets dependency

afg1 committed:
Use the correct probability from ONNX to match what sklearn expects

afg1 committed:
Print whole metrics dataframe in CLI

afg1 committed:
Use correct probability in gene classification

afg1 committed:
Switch to sha256 hash and include assembly id

This will change the names of everything we already calculated, so will need to re-run the whole pipeline

afg1 committed:
Force numeric output to match previous hash style

11 digits is the same as ensembl

afg1 committed:
Include genes in GFF export

This now makes the GFF export include genes as another object type in the gff file `predicted_gene` (alongside transcript and noncoding_exon).

For those transcripts that are part of a predicted gene, we also provide the parent predicted gene in the attributes, with the source noted as RNAcentral-gene-prediction

afg1 committed:
Exclude genes from bed export

afg1 committed:
Switch to pre-extracting region data for faster processing

afg1 committed:
Treat regions data as a path in preprocessing

afg1 committed:
Import pathlib

blakesweeney committed:
Rework gene search export

The search export for genes is done nearly entirely in rust. This
reworks it to match the new data structures internally. It also now
should export all the required fields for the real search export to
work. This is untested currently.

There are several other cleanups and tweaks as suggested by cargo
check/clippy.

afg1 committed:
Use ndjson for regions data

afg1 committed:
Increase classification batch size

afg1 committed:
Skip genes with mangled node IDs

afg1 committed:
Pull CPU specs from new SLURM environment variable

afg1 committed:
No longer writing species into additional fields tag.

It isn't required on the import side now, so some accessions don't have it, and more won't in the future. Stop writing it as an additional field so we don't have errors in the xml production

afg1 committed:
Memory optimization for feature preprocessing

Try doing everything in dataframes

Same modification applied in the training preprocessing side

blakesweeney committed:
Fix logic for gene building

This got the logic of validating the data for genes incorrect. I had to
negate the conditionals, rather than what I was doing. Secondly, I had
to make validating the member_count a warning. It seems there are
currently genes with inactive regions. As we need the data asap, I'm
making this a warning to look at later.

afg1 committed:
Memory optimisations for classification

Including:
- batched classification with random forest. Still fast (67M in 120s)
- full dataframe operations throughout, including stacking and rechunking
- caching of classification data to allow use of lazy API to do operations with data greater than the available RAM. Tidies up after itself
- Splitting the graph into subgraphs per chromosome

afg1 committed:
Update merge logic to only take gene members from new genes

Means the inactive ids are no longer needed, but leaving the argument there for temporary backwards compatibility

afg1 committed:
Stack features properly

afg1 committed:
Extract chromosome for low-data organisms

afg1 committed:
Use dataframe concatenation and fix some naming bugs

afg1 committed:
Merge branch 'genes-pipeline' into gene-search-export

afg1 committed:
Merge pull request #214 from RNAcentral/gene-search-export

Gene search export merge into genes pipeline

Adds the rust code needed to export new gene search terms

afg1 committed:
Optimise nextflow scheduling direcives based on last sucessful export

afg1 committed:
Update cargo lockfile

blakesweeney committed:
Rework genes and rust xml building

blakesweeney committed:
Prevent reverse strand hits

They will be opposite order, so this should fix that.

blakesweeney committed:
Start work to pass so_tree into gene building

blakesweeney committed:
Try to fix gene SO tree building

This should use the correct one, and the same one that sequence export
uses.

afg1 committed:
Dump the whole SO tree rather than only the tings in precompute

127 terms vs 118. This then doesn't choke on specific terms assigned by the genes type pipeline

afg1 committed:
Merge branch 'genes-pipeline' of github.com:RNAcentral/rnacentral-import-pipeline into genes-pipeline

blakesweeney committed:
Rework SO logic to track types and fix gene export

The gene export was putting out SO ids when it should have been SO
names. This fixes that so that names are exported, in order to make this
easier to track I added wrapper types for SO ids and SO names.

afg1 committed:
Use UInt32 to store the numeric part of the SO IDs

blakesweeney committed:
Assign all SO terms in tree to so_rna_type_name

afg1 committed:
Changes for rust pre-commit to make it more modular

afg1 committed:
Remove deprecated rustfmt option

afg1 committed:
Use custom rust format checker for precommit hook

afg1 committed:
Move some genes preprocessing to rust for speed

afg1 committed:
Massive rust formatting commit

Because pre-commit hooks are upset

afg1 committed:
Update poetry lockfile

afg1 committed:
Workaround for installing extension built by maturin without a virtualenv

afg1 committed:
Use optimised preprocessing code in gene feature calculation

afg1 committed:
Merge branch 'dev' into genes-pipeline

afg1 commented on .pre-commit-config.yaml:
```diff
@@ -16,12 +16,14 @@ repos:
         args: ["--profile", "black", "--filter-files"]
         name: isort (python)
--   repo: https://github.com/doublify/pre-commit-rust
-    rev: v1.0
+-   repo: local
     hooks:
-    -   id: fmt
-    -   id: cargo-check
-    -   id: clippy
```
These were removed because they were overzealous and caused me issues with irrelevant changes (I think it was the cargo edition changing and that leading to a warning about something, which these tools interpret as a failure)

afg1 commented on rnacentral_pipeline/rnacentral/genes/random_forest/convert.py:
```diff
@@ -0,0 +1,205 @@
+    return load_coordinates(path, taxid, regions_data)
+
+
+def database_to_bed(output_path: Path, taxid: int, conn_str: str) -> None:
```
We don't actually need this function anymore, since the genes are being exported in the GFF files, so I can remove it altogether

afg1 committed:
Add assembly id specification for BED export, and remove chromosome list to allow exporting all cheomosomes

afg1 committed:
Remove unused variables in batch inference loop

afg1 committed:
Remove unused numpy import

afg1 committed:
Remove unused re import

afg1 committed:
Remove unused time import

afg1 committed:
Remove unused base64 import

afg1 committed:
Remove unused imports

pmb59 reviewed (approved):
@afg1  LGTM, as this branch has been used successfully in production, but please consider the Gemini/copilot comments.

afg1 commented on files/ftp-export/rfam/rfam-annotations.sql:
```diff
@@ -14,5 +14,6 @@ JOIN rnc_rna_precomputed pre on pre.upi = hits.upi and taxid is null
 JOIN rfam_models models ON models.rfam_model_id = hits.rfam_model_id
 WHERE
     pre.is_active = true
+    AND hits.sequence_stop < hits.sequence_start
```
This is a good catch and I'm not sure why it has crept into the code this way around.

afg1 committed:
Fix Rfam export logic

hits.sequence_stop > hits.sequence_start means on the forward strand

hits.sequence_stop < hits.sequence_start from before would only export the reverse strand hits (which we have now deleted)

afg1 commented on utils/search-export/src/genes/region.rs:
```diff
@@ -140,15 +198,13 @@ where
-                assert!(false, "Bad data at {}", id);
-                None
+                panic!("Bad data at {}", id);
             },
```
@gemini-code-assist What error handling does thecaller actually have though? I don't mind if the application crashes because we want it to stop execution if there's an error in the data. If the caller does handle errors, maybe we can do something, but if not I'm inclined to leave it such that it will crash in a big way so we definitely can't ignore it

afg1 replied:
@gemini-code-assist I have made the proposed changes in commit [320b900](https://github.com/RNAcentral/rnacentral-import-pipeline/pull/221/commits/320b9003d84f650a6630b19b2e97ac30e0c1c416) can you have a look and see if it looks ok to you now?

afg1 committed:
Upgraded error handling in genes region iterator

afg1 commented on workflows/genes.nf:
```diff
@@ -0,0 +1,127 @@
+    path("so_embedding_model.emb")
+
+  """
+  wget -O so_embedding_model.emb ${params.genes.so_model_url}
+  """
```
I created a JIRA ticket for this, it would be a good enhancement
